config/frame_system.go

```diff
@@ -24,45 +27,89 @@ func (part *FrameSystemPart) ToProtobuf() (*pb.FrameSystemConfig, error) {
 	}
-	poseInFrame := &commonpb.PoseInFrame{
-		ReferenceFrame: part.FrameConfig.Parent,
-		Pose:           spatialmath.PoseToProtobuf(part.FrameConfig.Pose()),
+	pose, err := part.FrameConfig.Pose()
+	if err != nil {
+		return nil, err
 	}
-	var modelJSON []byte
-	var err error
+	
+	geom, err := part.FrameConfig.Geometry.ToProtobuf()
+	if err != nil {
+		return nil, err
+	}
+	linkFrame := &commonpb.StaticFrame{
+		Name: part.FrameConfig.ID,
+		PoseInParentFrame: &commonpb.PoseInFrame{
+			ReferenceFrame: part.FrameConfig.Parent,
+			Pose:           spatialmath.PoseToProtobuf(pose),
+		},
+		Geometries: []*commonpb.Geometry{geom},
+	}
+	var modelJson map[string]interface{}
 	if part.ModelFrame != nil {
-		modelJSON, err = part.ModelFrame.MarshalJSON()
+		bytes, err := part.ModelFrame.MarshalJSON()
 		if err != nil {
 			return nil, err
 		}
+		json.Unmarshal(bytes, &modelJson)
+	}
+	kinematics, err := protoutils.StructToStructPb(modelJson)
+	if err != nil {
+		return nil, err
 	}
 	return &pb.FrameSystemConfig{
-		Name:              part.Name,
-		PoseInParentFrame: poseInFrame,
-		ModelJson:         modelJSON,
+		Frame: linkFrame,
+		Kinematics:         kinematics,
 	}, nil
 }
 
 // ProtobufToFrameSystemPart takes a protobuf object and transforms it into a FrameSystemPart.
 func ProtobufToFrameSystemPart(fsc *pb.FrameSystemConfig) (*FrameSystemPart, error) {
-	pose := spatialmath.NewPoseFromProtobuf(fsc.PoseInParentFrame.Pose)
-	frameConfig := &Frame{
-		Parent:      fsc.PoseInParentFrame.ReferenceFrame,
-		Translation: pose.Point(),
-		Orientation: pose.Orientation(),
+	pose := spatialmath.NewPoseFromProtobuf(fsc.Frame.PoseInParentFrame.Pose)
+	orient, err := spatialmath.NewOrientationConfig(pose.Orientation())
+	if err != nil {
+		return nil, err
+	}
+	var geom *spatialmath.GeometryConfig
+	if len(fsc.Frame.Geometries) > 0 {
+		geomTemp, err := spatialmath.NewGeometryCreatorFromProto(fsc.Frame.Geometries[0])
+		if err != nil {
+			return nil, err
+		}
+		geom, err = spatialmath.NewGeometryConfig(geomTemp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	frameConfig := &referenceframe.LinkCfg{
+		&referenceframe.StaticFrameCfg{
+			ID: fsc.Frame.Name,
+			Translation: *spatialmath.NewTranslationConfig(pose.Point()),
+			Orientation: orient,
+			Geometry: geom,
+		},
+		fsc.Frame.PoseInParentFrame.ReferenceFrame,
 	}
 	part := &FrameSystemPart{
-		Name:        fsc.Name,
 		FrameConfig: frameConfig,
 	}
-	modelFrame, err := referenceframe.UnmarshalModelJSON(fsc.ModelJson, fsc.Name)
-	if err != nil {
-		if errors.Is(err, referenceframe.ErrNoModelInformation) {
-			return part, nil
+	fmt.Println("id", frameConfig.ID)
+	fmt.Println("fsc", fsc.Frame)
+	fmt.Println("part", part.FrameConfig)
+	if len(fsc.Kinematics.AsMap()) > 0 {
+		modelBytes, err := json.Marshal(fsc.Kinematics.AsMap())
+		if err != nil {
+			return nil, err
 		}
-		return nil, err
+		modelFrame, err := referenceframe.UnmarshalModelJSON(modelBytes, fsc.Frame.Name)
```

Comment:
This is what we touched on briefly in the meeting with Eliot; swapping the proto field from `bytes` to a `Struct` comes with over-the-wire communication benefits, but with the downside of needing additional processing at the other end.

Re-interpreting a generic `Struct` as json bytes should be super inexpensive, fortunately.

Since the "official" entry point for parsing these structs is starting with json, that's probably the best way to have this, since we don't want to have multiple parsers for the format.

Reply:
Ah, thanks for explaining!